### PR TITLE
[REVIEW] Use dask_cudf partitions for iteration and support DataFrame-based Dataset input

### LINKED ISSUE #163: [FEA] Accept cudf and dask_cudf DataFrame objects in the `Dataset` API
**Is your feature request related to a problem? Please describe.**
NVTabular does not currently support transformations that result in the global shuffling of data (prior to the writing of data).  Rather than supporting a dedicated phase for shuffling,  I suggest that the `Dataset` API be expanded to accept an existing `dask_cudf.DataFrame` (or `cudf.DataFrame`) object.  This would allow users to perform arbitrary `dask_cudf` operations outside the constraints of `nvtabular.Workflow`.

**Describe the solution you'd like**
I would like the following logic to work:

```python
import nvtabular as nvt
import nvtabular.ops as ops
import dask_cudf

# Create dask_cudf DataFrame and perform arbitrary operation(s)
ddf = dask_cudf.read_parquet(<path>)
ddf = ddf.sort_values("user_id")

# Crate an nvt Dataset object from our `ddf` (NOT from paths)
dataset = nvt.Dataset(ddf)

# Define Workflow, and apply as usual
wf = nvt.Workflow(...)
wf.add_feature([ops.ZeroFill(), ops.LogOp()])
wf.apply(dataset, ...)
```

**Describe alternatives you've considered**
The alternatives is to require users to persist transformed data to disk before defining their `Dataset` object.  Although this may be "best" in some cases, it is certainly far from ideal when IO is expensive, or when the dataset fits comfortably in global device memory.

[**EDIT**: Note that an earlier version of this issue suggested that we accept non-Dataset types to `Workflow.apply` - I think it is **much** easier to handle this within `Dataset` (so that we can do things like `dataset.to_iter()` in our workflow)]

### PR DESCRIPTION
**Pull request by rjzamora**
Closes #163 - Users can now pass in a pandas/cudf/dask_cudf DataFrame object to the `Dataset` API.  For now, we use a single partition for pandas/cudf, and preserve the number of partitions for dask_cudf. In the future, we can use `part_mem_fraction` to repartition things differently.

Commit `b25ab30` also removes `GPUFileReader`, `PQFileReader`, `CSVFileReader`, and `GPUFileIterator` -- I needed to change some dataloader code to make this work.

### CODE REVIEW

**Comment by rjzamora** [comment]
@benfred @jperez999 @alecgunny -  Let me know if we need to hold off on yanking the iterator classes.  While this is definitely something we need to do eventually,   I do want to avoid unnecessary dataloader headaches.  For example, the `Dataset` API does not currently support partitioning on a specific row-count (because it is considered a bad idea for parquet - which has been the focus of development).  This means, the dataloaders will not be able to iterate through the dataset using a row-based `batch_size` (they will need to rely on `part_mem_fraction`, or `part_size`).

**Comment by jperez999** [comment]
I have no problems with ripping out the old iteration methodology.

**Comment by rjzamora** [comment]
> I have no problems with ripping out the old iteration methodology.

Thanks @jperez999 - It would be great to get your eyes on the dataloader changes if/when you get some free time :)

**Comment by rjzamora** [comment]
Just wanted to motivate this PR a bit further by pointing out that this allows us to (indriectly) handle a `drop_duplicates` operation.  For example, the user can use either dask.dataframe or dask_cudf to create an initial `Dataset` object that is partitioned by the subset of column(s) with unwanted duplicates.  For there, a trivial `LocalDropDuplicates` Op can be used to drop local duplicates...

```python
import dask_cudf
import dask
import nvtabular as nvt

# Create a `Dataset` with all unique "id" values
# in same chunk/partition
def _get_shuffled_ddf():
    ddf = dask.datasets.timeseries()
    ddf = dask_cudf.from_dask_dataframe(ddf)
    return ddf.shuffle("id", ignore_index=True)
ds = nvt.Dataset(_get_shuffled_ddf())

# User-defined Op to drop duplicates within a chunk/partition
class LocalDropDuplicates(nvt.ops.TransformOperator):
    default_in, default_out = "all", "all"
    def __init__(self, subset=None, **kwargs):
        super().__init__(**kwargs)
        self.subset = subset
    def op_logic(self, gdf, target_columns, stats_context=None):
        return gdf[target_columns].drop_duplicates(subset=self.subset)

# Define/apply Workflow
cat_names = ["id"]
cont_names = ["x", "y"]
label_name = ["name"]
wf = nvt.Workflow(cat_names=cat_names, cont_names=cont_names, label_name=label_name)
wf.add_preprocess(nvt.ops.Normalize())
wf.add_preprocess(LocalDropDuplicates(subset=["id"]))
wf.add_preprocess(nvt.ops.Categorify(cat_cache="device", tree_width=1))
wf.finalize()
wf.apply(ds, output_format=None)

# View head of some partition...
print(wf.get_ddf().get_partition(1).head())
```
Output:
```
           x         y   id      name
28  1.102895  0.983658   68    Victor
3   1.419708  1.576253   89  Patricia
0  -0.536301  1.624180  172     Frank
2   1.233830  1.209515  183     Frank
7   1.427104  0.486176  194     Alice
```

File: nvtabular/torch_dataloader.py [hunk 1]
```diff
@@ -210,9 +203,9 @@ def __init__(
         self.cat_cols = cats
         self.cont_cols = conts
         self.label_cols = labels
-        self.itr = GPUFileIterator(path, **kwargs)
+        self.itr = Dataset(path, **kwargs).to_iter(columns=cats + conts + labels)
```
[the remark this answers is not in the document]
Comment (benfred) [reply nvtabular/torch_dataloader.py:206]: I've created https://github.com/NVIDIA/NVTabular/issues/168 to track this

**Pull request merged**